Rando: More Misc. Hints

stratomaster64 opened the pull request:
Adds:
- Saria as an NPC to receive a hint on where the progressive magic meter would be.
  - The ability to utilize Saria's song to get this hint if you don't see her at her usual spots.
- Hint for frog ocarina game by stepping on the platform near them in Zora's River.
- Sheik as an NPC to receive the light arrows hint, when trials are required.
  - You can see her in the Temple of Time and Ganon's Castle as adult.
  - Ganondorf will only say the funny line, but this should allow light arrow hint to only need the rainbow bridge.
 - A function within `randomizer.cpp` to allow for context-sensitive message replacement...I'm guessing? It just ended up that way.
Hopefully we should hit hint parity once this makes it in.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/927719245.zip)
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/927719246.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/927719247.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/927719248.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/927719249.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/927719250.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/927719251.zip)
<!--- section:artifacts:end -->

stratomaster64 committed:
commit constipation

stratomaster64 committed:
Merge branch 'develop' into yet-more-hints

etreff commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -4873,6 +4910,64 @@ CustomMessage Randomizer::GetWarpSongMessage(u16 textId, bool mysterious) {
+                if (originalTextId == TEXT_SHEIK_1) {
+                    messageEntry.Replace("{{message}}", 
+                    "@,&meet me at %gGanon's Castle%w&once you obtain the %rkey to his lair%w.",
+                    "Test",
```
Are these Test strings supposed to be here?

stratomaster64 replied:
They're placeholders for French and German translations. Normally I would use AI-generated (read: DeepL) translations but I just forgot 😂

stratomaster64 committed:
fix inconsistencies between 3d and soh settings

stratomaster64 committed:
Add RSK check for Saria; Sheik now tells you other reqs for ganon

stratomaster64 committed:
Translations (thanks Purple and Timmy_GamerNepgear);
Retain Saria's "face-to-face" text in rando if you're too close

stratomaster64 committed:
Merge branch 'develop' into yet-more-hints

Commit:
Merge branch 'HarbourMasters:develop' into yet-more-hints

Commit:
Merge branch 'HarbourMasters:develop' into yet-more-hints

Commit:
Merge branch 'HarbourMasters:develop' into yet-more-hints

Commit:
Merge branch 'HarbourMasters:develop' into yet-more-hints

stratomaster64 commented:
May want to make this PR dependent on #2981 just so I can handle Sheik's MS hint text all in one go, rather than make a follow-up PR.

Commit:
Merge branch 'HarbourMasters:develop' into yet-more-hints

Commit:
Merge branch 'HarbourMasters:develop' into yet-more-hints

Commit:
Merge branch 'develop' into yet-more-hints

Commit:
Merge branch 'HarbourMasters:develop' into yet-more-hints

Commit:
Merge branch 'develop' into yet-more-hints

stratomaster64 committed:
fix scene renames for sheik

stratomaster64 committed:
whoops

stratomaster64 committed:
Undo LA hint setting erasure from conflict

aMannus commented on soh/soh/Enhancements/randomizer/3drando/hints.cpp:
```diff
@@ -820,19 +830,50 @@ void CreateGregRupeeHint() {
 
   Text temp1 = Text{
     "By the way, if you're interested, I saw the shiniest %gGreen Rupee%w somewhere in%g ",
-    "",
+    "Au fait, si ça t'intéresse, j'ai aperçu le plus éclatant des %gRubis Verts%w quelque part à %g",
     ""
   };
 
   Text temp2 = {
     "%w.^It's said to have %rmysterious powers%w...^But then, it could just be another regular rupee.&Oh well.",
-    "",
+    "%w. On dit qu'il possède des pouvoirs mystérieux... Mais bon, ça pourrait juste être un autre rubis ordinaire.",
```
Looks like this is a translation for the Greg hint, was this intended to be in this PR? Is this AI generated or properly translated?

stratomaster64 replied:
This was properly translated by Hato, and was included here since the translation was missed on the original Greg PR.

aMannus commented on soh/soh/Enhancements/randomizer/3drando/hints.cpp:
```diff
@@ -820,19 +830,50 @@ void CreateGregRupeeHint() {
+  //Get the location of the light arrows
+  auto lightArrowLocation = FilterFromPool(allLocations, [](const uint32_t loc){return Location(loc)->GetPlaceduint32_t() == LIGHT_ARROWS;})[0];
+  Text area = GetHintRegion(Location(lightArrowLocation)->GetParentRegionKey())->GetHint().GetText();
+  Text temp1 = Text{
+    "I overheard Ganondorf say that he misplaced the %rLight Arrows%w in&%g",
+    "J'ai entendu dire que Ganondorf aurait caché les %rFlèches de Lumière%w dans %g",
+    ""
+  };
```
Is this AI generated or made by a human? If this is AI translated, could we ask our resident French translator @PurpleHato or someone else from the community to double check all the French text in the PR?

Also, I assume the german is missing because 3DS rando logic doesn't actually accept German? Might be something to note for a potential rando refactor to either implement german or at the very least make it so we don't need empty strings for German.

stratomaster64 replied:
This was translated by Hato. And yes, our 3d rando doesn't currently support German, opting for Spanish instead.

Malkierian replied:
Made note of the language issue in our rando refactor idea tracker.

PurpleHato replied:
> Is this AI generated or made by a human? If this is AI translated, could we ask our resident French translator @PurpleHato or someone else from the community to double check all the French text in the PR?
> 
> Also, I assume the german is missing because 3DS rando logic doesn't actually accept German? Might be something to note for a potential rando refactor to either implement german or at the very least make it so we don't need empty strings for German.

Those were made by me, and I completely forgot to ask for the German translations, they could be a follow-up PR from my side since I do not know when our Timmy will be able to translate them once I give them to him and I do not want that to be a blocker

Side note for myself: Also need to add the 3DS system for dungeon prefix for French and German on everything that name a place, since we need to say for instance "l'Arbre Mojo", "Le Temple du Feu", "La Forêt Kokiri".

aMannus commented on soh/soh/Enhancements/randomizer/3drando/hints.cpp:
```diff
@@ -820,19 +830,50 @@ void CreateGregRupeeHint() {
+  auto magicLocation = FilterFromPool(allLocations, [](const uint32_t loc){return Location(loc)->GetPlaceduint32_t() == PROGRESSIVE_MAGIC_METER;})[0];
+  Text area = GetHintRegion(Location(magicLocation)->GetParentRegionKey())->GetHint().GetText();
+  Text temp1 = Text{
+    "Did you feel the %gsurge of magic%w recently? A mysterious bird told me it came from %g",
+    "As-tu récemment ressenti une vague de %gpuissance magique%w? Un mystérieux hibou m'a dit  qu'elle provenait du %g",
+    ""
+  };
+  Text temp2 = Text{
+    "%w.^You should check that place out, @!$C", 
+    "%w. Tu devrais aller y jeter un coup d'oeil, @!", 
+    "%w."
```
Same here, is this AI translated?

stratomaster64 replied:
Translated by Hato.

aMannus commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -1265,9 +1294,6 @@ std::string FormatJsonHintText(std::string jsonHint) {
             formattedHintMessage.replace(start_pos, textToReplace.length(), iconString);
         }
     }
-
-    formattedHintMessage += 0x02;
-
     return formattedHintMessage;
```
Just curious, why is this no longer needed?

stratomaster64 replied:
When accessing Path of Magic hints from using Saria's song, a certain control character (specifically `0x14`) is used to delegate control to the game so Navi can talk to you again and you can close the textbox when ready. Adding that character breaks that flow and causes a softlock.

aMannus replied:
My worry is more about why this was here before. Does removing this not break anything else?

stratomaster64 replied:
From what I could tell, the game itself adds a `0x02` to signify EOF for messages. So most custom messages before had duplicate endings which made this safe to remove.

aMannus commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -4439,12 +4477,16 @@ void RandomizerSettingsWindow::DrawElement() {
                 UIWidgets::InsertHelpHoverText("Talking to the chest game owner after buying a key will tell you the location of Greg the Green Rupee.");
+                UIWidgets::PaddedEnhancementCheckbox("Saria (Magic)", "gRandomizeSariaHint", true, false);
+                UIWidgets::InsertHelpHoverText("Talking to Saria either in person or through Saria's Song will tell you the location of a progressive magic meter.");
+                UIWidgets::PaddedEnhancementCheckbox("Frog Ocarina Game", "gRandomizeFrogsHint", true, false);
+                UIWidgets::InsertHelpHoverText("Standing near the frogs in Zora's River will tell you the reward for the frogs' ocarina game.");
```
What does "standing near" mean?

stratomaster64 replied:
When you are within proximity of the frogs as child (don't remember how many units), there's a message that pops up. Usually, you'll be close to the platform to play for them.

aMannus replied:
Appreciate the more specific description in the last commit too!

aMannus commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -441,11 +443,23 @@ void Randomizer::LoadHintLocations(const char* spoilerFileName) {
         Randomizer::hintMessageTableID, TEXT_GANONDORF_NOHINT,
         CustomMessage(gSaveContext.ganonText, gSaveContext.ganonText, gSaveContext.ganonText));
+    CustomMessageManager::Instance->CreateMessage(
+        Randomizer::hintMessageTableID, TEXT_SHEIK_1,
+        CustomMessage("{{message}}", "{{message}}", "{{message}}"));
+    CustomMessageManager::Instance->CreateMessage(
+        Randomizer::hintMessageTableID, TEXT_SHEIK_2,
```
What does TEXT_SHEIK_1 and TEXT_SHEIK_2 stand for? Could we name these more specifically?

stratomaster64 replied:
TEXT_SHEIK_1 is text ID 0x700F, which triggers when talking to Sheik while you don't have the hookshot, and reads as follows according to CM wiki:

> To save the forest girl, you need another skill...Head to Kakariko Village!

TEXT_SHEIK_2 is text ID 0x7010, which triggers when talking to Sheik after getting hookshot but before clearing Forest Temple, and reads as follows:

> It looks like you have the skills you need...The forest girl is waiting for your help...

Name suggestions are welcome.

aMannus replied:
I think the names you choose yourself are fine.

aMannus reviewed (commented):
Initial quick scan from me. Looks like it follows the same pattern as many other existing hints, so that's great. I do have some questions.

Malkierian commented:
Other than what's been mentioned already by others, the code looks good to me. If I get around to actually playtesting it, I'll drop an approving review.

stratomaster64 committed:
Clarified Sheik text IDs; Clarified final frogs hint

stratomaster64 committed:
Fixed Sheik text ID assignment; Fixed Saria's messages in `OTRGlobals`

stratomaster64 committed:
Added hint locs for sheik and saria

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -942,6 +969,7 @@ void Randomizer::ParseRandomizerSettingsFile(const char* spoilerFileName) {
                         } else if(it.value() == "Minimal") {
                             gSaveContext.randoSettings[index].value = RO_ITEM_POOL_MINIMAL;
                         }
+                        break;
```
ermm,, nice find. I wonder if this has been causing any issues 😆

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -5229,6 +5274,70 @@ CustomMessage Randomizer::GetWarpSongMessage(u16 textId, bool mysterious) {
     return messageEntry;
 }
 
+CustomMessage Randomizer::GetMiscMessage(s16 scene, u16 originalTextId) {
```
I don't see much benefit to this being one function that splits into 3 blocks vs just 3 different functions personally. And the scene argument only seems relevant for the middle block

That being said, I think a lot of these custom message handlers could be cleaned up so up to you if you want to make the switch, I think this system will look a bit different for v3/v4 :tm:

briaguya0 replied:
i also think splitting it into 3 makes sense
`GetFrogMessage`
`GetSheikMessage`
`GetSariaMessage`

we're already making calls for each from separate places in `CustomMessage_RetrieveIfExists`, so we might as well call into different methods.

it could be cool to see a bunch of the logic in `CustomMessage_RetrieveIfExists` moved out and cleaned up, but that feels like it should be its own PR

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -5229,6 +5274,70 @@ CustomMessage Randomizer::GetWarpSongMessage(u16 textId, bool mysterious) {
+                break;
+        }
+        return messageEntry;
+    } else {
```
If this function does end up sticking around, this should be else if'd to match the saria messages, in prep for people continue adding more handled messages to this

garrettjoecox commented on soh/soh/SaveManager.cpp:
```diff
@@ -218,6 +218,12 @@ void SaveManager::LoadRandomizerVersion2() {
     std::string gregHintText;
     SaveManager::Instance->LoadData("gregHintText", gregHintText);
     memcpy(gSaveContext.gregHintText, gregHintText.c_str(), gregHintText.length());
+    std::string sheikText;
+    SaveManager::Instance->LoadData("sheikText", sheikText);
+    memcpy(gSaveContext.sheikText, sheikText.c_str(), sheikText.length());
+    std::string sariaText;
+    SaveManager::Instance->LoadData("sariaText", sariaText);
+    memcpy(gSaveContext.sariaText, sariaText.c_str(), sariaText.length());
```
There's actually a bug here with other hints we need to fix, we need to copy the termination char over (someone can spot check me on this)
```suggestion
    std::string sheikText;
    SaveManager::Instance->LoadData("sheikText", sheikText);
    memcpy(gSaveContext.sheikText, sheikText.c_str(), sheikText.length() + 1);
    std::string sariaText;
    SaveManager::Instance->LoadData("sariaText", sariaText);
    memcpy(gSaveContext.sariaText, sariaText.c_str(), sariaText.length() + 1);
```

garrettjoecox commented on soh/src/code/z_play.c:
```diff
@@ -745,6 +745,14 @@ void Play_Init(GameState* thisx) {
                     GET_PLAYER(play)->actor.world.pos.y + Player_GetHeight(GET_PLAYER(play)) + 5.0f,
                     GET_PLAYER(play)->actor.world.pos.z, 0, 0, 0, 1, true);
     }
+    if (gSaveContext.n64ddFlag && LINK_IS_ADULT &&
```
Can we move this to a hook? If not I'd either extract this to a clearly named function or make a comment explaining what it's doing. Obvious to us now but 6 months from now "why are we spawning en_xc here?"

stratomaster64 replied:
I don't have the most experience with hooks or how they work here, but I'd be willing to move this to one.

garrettjoecox reviewed (commented):
A few comments, solid wooooork

stratomaster64 committed:
Set up hook for Sheik spawn;
Sheik no longer cheats with room transitions;
Enforced text IDs on saria msg function

stratomaster64 committed:
Set up hook for Sheik spawn;
Sheik no longer cheats with room transitions;
Enforced text IDs on saria msg function

stratomaster64 committed:
Update soh/soh/SaveManager.cpp

Co-authored-by: Garrett Cox <garrettjcox@gmail.com>

stratomaster64 committed:
Merge branch 'yet-more-hints' of https://github.com/stratomaster64/Shipwright into yet-more-hints

garrettjoecox commented on soh/soh/Enhancements/mods.cpp:
```diff
@@ -953,6 +953,23 @@ void RegisterAltTrapTypes() {
     });
 }
 
+void RegisterSheikSpawn() {
```
```suggestion
void RegisterRandomizerSheikSpawn() {
```
or
```suggestion
void RegisterSheikSpawnForRandomizer() {
```

garrettjoecox reviewed (approved):
Noice.

Commit:
nice.gif

Co-authored-by: Garrett Cox <garrettjcox@gmail.com>

stratomaster64 committed:
got ahead of myself

briaguya0 commented on soh/soh/Enhancements/mods.cpp:
```diff
@@ -953,6 +953,23 @@ void RegisterAltTrapTypes() {
+void RegisterRandomizerSheikSpawn() {
+    GameInteractor::Instance->RegisterGameHook<GameInteractor::OnSceneSpawnActors>([]() {
+        if (!gPlayState) return;
+        bool canSheik = OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_TRIAL_COUNT) != RO_GANONS_TRIALS_SKIP;
```
does it make sense to spawn sheik even when the light arrow hint is turned off? i think i'd vote to have sheik only appear when the setting is on.

stratomaster64 replied:
It should be a combination of the two, as Sheik relies on LA hint *and* there being trials (otherwise you go to Ganondorf for LA hint).

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -2041,11 +2041,14 @@ extern "C" int CustomMessage_RetrieveIfExists(PlayState* play) {
                : CustomMessageManager::Instance->RetrieveMessage(Randomizer::hintMessageTableID, TEXT_ALTAR_CHILD);
         } else if (textId == TEXT_GANONDORF) {
-            if (INV_CONTENT(ITEM_ARROW_LIGHT) == ITEM_ARROW_LIGHT || !Randomizer_GetSettingValue(RSK_GANONDORF_LIGHT_ARROWS_HINT)) {
+            if (INV_CONTENT(ITEM_ARROW_LIGHT) == ITEM_ARROW_LIGHT || !Randomizer_GetSettingValue(RSK_LIGHT_ARROWS_HINT) ||
+                Randomizer_GetSettingValue(RSK_LIGHT_ARROWS_HINT) && Randomizer_GetSettingValue(RSK_GANONS_TRIALS)) {
```
i don't think we need this change. the point of light arrow hint outside of trials is people need the lights to get past the trials. it would be out of logic to get to dorf without them, and i guess i see no reason to not give the light arrow hint to someone who decides to trial skip past the sheik hint without lights.

Commit:
Merge branch 'HarbourMasters:develop' into yet-more-hints

briaguya0 commented on soh/src/overlays/actors/ovl_En_Xc/z_en_xc.c:
```diff
@@ -2224,11 +2239,27 @@ void EnXc_SetupDialogueAction(EnXc* this, PlayState* play) {
+         this->actor.flags |= ACTOR_FLAG_TARGETABLE | ACTOR_FLAG_FRIENDLY;
+        if (gSaveContext.n64ddFlag && gPlayState->sceneNum == SCENE_TEMPLE_OF_TIME) {
+            if (!CHECK_DUNGEON_ITEM(DUNGEON_KEY_BOSS, SCENE_GANONS_TOWER)) {
+                this->actor.textId = 0x700F;    
```
it'd be nice to use `#define`s or an enum for these instead of having them as magic numbers

stratomaster64 replied:
I agree, although this file had the original text IDs as magic numbers.

briaguya0 commented on soh/src/overlays/actors/ovl_En_Xc/z_en_xc.c:
```diff
@@ -2401,6 +2432,9 @@ void EnXc_Init(Actor* thisx, PlayState* play) {
         case SHEIK_TYPE_0:
             EnXc_DoNothing(this, play);
             break;
+        case -1: //Special rando case
```
let's make this not a magic number

briaguya0 reviewed (commented):
overall this is looking really good! couple little comments in there but nothing major.

Archez commented on soh/soh/Enhancements/mods.cpp:
```diff
@@ -953,6 +953,23 @@ void RegisterAltTrapTypes() {
+        bool canSheik = OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_TRIAL_COUNT) != RO_GANONS_TRIALS_SKIP;
+        if (!gSaveContext.n64ddFlag || !LINK_IS_ADULT || !canSheik) return;
+        switch (gPlayState->sceneNum) {
+            case SCENE_TEMPLE_OF_TIME:
```
You are only looking at the scene, and not the room that you want to spawn Sheik in. This means that every room transition e.g. Ganon's tower (going to all the trials) is going to spawn Sheik, even though you cant get to her.

You should be checking for room here too before spawning.

stratomaster64 committed:
Conquered and divided; simplified stuff

garrettjoecox commented on soh/src/overlays/actors/ovl_En_Xc/z_en_xc.h:
```diff
@@ -10,6 +10,7 @@ typedef void (*EnXcActionFunc)(struct EnXc*, PlayState*);
 typedef void (*EnXcDrawFunc)(struct Actor*, PlayState*);
 
 typedef enum {
+    /*-1 */ SHEIK_TYPE_RANDO = -1,
```
I don't think this really makes any real difference, but typically when I've added to source enums I'll add it to the end. Not sure if one is preferable over the other

stratomaster64 replied:
I only really placed it there because of the number ordering.

briaguya0 replied:
what's the reasoning behind using `-1` instead of `10` here? just curious

stratomaster64 replied:
Weird that I have to do it from a specific tab in GH (thanks Microsoft), but see the comment below.

briaguya0 replied:
It's possible for the type to be something not in the enum with vanilla play but not negative?

stratomaster64 replied:
I don't think so, but my reasoning was more along the lines of "this is intentionally meant for randomizer." Either -1 or 10 wouldn't matter anyway, as Sheik will kill her actor if the rando enum is used outside randomizer anyway.

Malkierian commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -5229,6 +5274,73 @@ CustomMessage Randomizer::GetWarpSongMessage(u16 textId, bool mysterious) {
+CustomMessage Randomizer::GetSariaMessage(u16 originalTextId) {
+    if (originalTextId == TEXT_SARIA_SFM || originalTextId == TEXT_SARIAS_SONG_FOREST_SOUNDS || TEXT_SARIAS_SONG_FOREST_TEMPLE) {
+        CustomMessage messageEntry = CustomMessageManager::Instance->RetrieveMessage(Randomizer::hintMessageTableID, TEXT_SARIAS_SONG_FACE_TO_FACE);
+        CustomMessage messageEntry2 = messageEntry;
```
Why do you instantiate a second messageEntry here? Why not operate on the first one in the rest of this function?

stratomaster64 replied:
I remember having issues replacing the control character with the first custom message...maybe that's not the case anymore?

briaguya0 commented on soh/soh/Enhancements/randomizer/3drando/hints.cpp:
```diff
@@ -192,14 +203,18 @@ Text& GetWarpPreludeText() {
 
-std::string GetGanonHintLoc() {
-    return ganonHintLoc;
+std::string GetlightArrowHintLoc() {
```
nit
```suggestion
std::string GetLightArrowHintLoc() {
```

briaguya0 commented on soh/soh/Enhancements/randomizer/3drando/hints.hpp:
```diff
@@ -224,4 +228,5 @@ Text& GetWarpNocturneText();
 
 std::string GetDampeHintLoc();
-std::string GetGanonHintLoc();
+std::string GetlightArrowHintLoc();
```
nit
```suggestion
std::string GetLightArrowHintLoc();
```

briaguya0 commented on soh/src/code/z_play.c:
```diff
@@ -745,6 +745,7 @@ void Play_Init(GameState* thisx) {
                     GET_PLAYER(play)->actor.world.pos.y + Player_GetHeight(GET_PLAYER(play)) + 5.0f,
                     GET_PLAYER(play)->actor.world.pos.z, 0, 0, 0, 1, true);
     }
+
```
```suggestion
```

briaguya0 reviewed (approved):
there are still a couple magic numbers in there (specifically in saria) but that can be a part of a bigger magic number cleanup effort later (and not part of this PR)

couple tiny naming/formatting nits but nothing blocking. i say we :shipit:

stratomaster64 commented:
@briaguya-ai For some reason I'm not able to add to or reply to the comment regarding the rando type on Sheik. The theory was that the specific would never occur in gameplay unless intentionally set.

Commit:
Merge branch 'develop' into yet-more-hints

Commit:
nitpicking

Co-authored-by: briaguya <70942617+briaguya-ai@users.noreply.github.com>

Commit:
Apply suggestions from code review

Co-authored-by: briaguya <70942617+briaguya-ai@users.noreply.github.com>

stratomaster64 committed:
the little things i stg

aMannus reviewed (approved):
I'd like to see the duplicate messageEntry that Malk mentioned be removed, and the -1 enum be changed, but those are both minor and not dealbreakers to me.

Nicely done 👍
